### ISSUE TITLE
Existence of POM file in one repository requires all artifacts to be in the same repository

### ISSUE THREAD

**Issue by mnonnenmacher** [gradle/gradle#3666]
### Expected Behavior

If one repository does contain a POM file for an artifact, but not the requested artifact, Gradle is checking the next repository for its existence.

### Current Behavior

Gradle stops at the first repository that contains the POM file.

### Context

When an artifact is present in only one of the repositories it should not be required to change the order of the repositories so that it can be resolved by Gradle. This is even worse in a circular situation as described below.

### Steps to Reproduce (for bugs)

Example in: https://github.com/mnonnenmacher/gradle-repository-issue

To demonstrate this check this build.gradle:

```gradle
apply plugin: 'java'

repositories {
    jcenter()

    maven {
        url 'https://repo.jenkins-ci.org/releases/'
    }
}

dependencies {
    compile 'org.jenkins-ci.main:jenkins-test-harness:2.32'
    compile 'org.jenkins-ci.main:jenkins-war:2.92'
}
```

Running `./gradlew` build fails with:

```bash
Could not resolve all files for configuration ':compileClasspath'.
> Could not find bootstrap-core-assets.jar (org.jenkins-ci.ui:bootstrap:1.3.2).
  Searched in the following locations:
      https://jcenter.bintray.com/org/jenkins-ci/ui/bootstrap/1.3.2/bootstrap-1.3.2-core-assets.jar
```

Because JCenter does contain the POM file for the bootstrap dependency, but not the core-assets.jar, which is present in the jenkins-ci repo.

https://jcenter.bintray.com/org/jenkins-ci/ui/bootstrap/1.3.2/
https://repo.jenkins-ci.org/releases/org/jenkins-ci/ui/bootstrap/1.3.2/

A similar problem occurs when switching the order of the repositories to:

```gradle
repositories {
    maven {
        url 'https://repo.jenkins-ci.org/releases/'
    }

    jcenter()
}
```

This time `./gradlew build` fails with:

```bash
Could not resolve all files for configuration ':compileClasspath'.
> Could not find org-netbeans-insane.jar (org.netbeans.modules:org-netbeans-insane:RELEASE72).
  Searched in the following locations:
      https://repo.jenkins-ci.org/releases/org/netbeans/modules/org-netbeans-insane/RELEASE72/org-netbeans-insane-RELEASE72.jar
```

Because the jenkins-ci repo does not contain the org-netbeans-insane JAR file which is present in JCenter.

https://jcenter.bintray.com/org/netbeans/modules/org-netbeans-insane/RELEASE72/
https://repo.jenkins-ci.org/releases/org/netbeans/modules/org-netbeans-insane/RELEASE72/

This leads to a situation where it is impossible to build the project without exluding one of the transitive dependencies, even when all required artifacts are available in the repositories.

This bug seems to be related:
https://issues.gradle.org/browse/GRADLE-2034

### Your Environment

 * Build scan URL: Did not run a build scan.

**Comment by sschuberth** [comment]
I wonder whether this is resolved already in Gradle 4.4 by the [plugin resolution changes](https://docs.gradle.org/4.4/release-notes.html#plugin-repositories-enhancements):

> Plugin resolution now takes all plugin repositories into account and can resolve transitive plugin dependencies across them. Previous versions of Gradle required that all transitive dependencies of a given plugin were present in the same repository as the plugin.

**Comment by mnonnenmacher** [comment]
@sschuberth: I have upgraded the sample project to Gradle 4.4 but it still fails for the same reason.

**Comment by bigdaz** [comment]
Gradle considers a `repository` to provide either a complete module, or none at all. There is no support for a single module that is spread across multiple repositories. So if Gradle finds a POM file for a particular module in a repository, it will then only look in that repository for artifacts of that module.

You can, however, defined additional `artifactUrls` for a repository. Gradle will then search these additional locations for module.

However, in this case it seems like https://repo.jenkins-ci.org/releases/org/netbeans/modules/org-netbeans-insane is not a valid/complete Maven module.
There are a couple of ways we could address this:

1. Implement #1369, so that you can tell Gradle not to look for `org-netbeans-insane` in the https://repo.jenkins-ci.org/releases repository.
2. Verify the existence of the primary jar file whenever a POM file is resolved, and ignore the module if a POM without artifacts is discovered. Unfortunately this will result in additional network requests when the artifact is not actually required. (I've opened #3897 for this).

I'm closing this issue in favour of the above linked issues.

**Comment by mnonnenmacher** [comment]
Thanks for the explanation @bigdaz. For reference there is an open bug in the Jenkins JIRA about the inconsistent org.netbeans.modules:org-netbeans-insane:RELEASE72 artifact:
https://issues.jenkins-ci.org/browse/INFRA-900

**Comment by Ordiel** [comment]
Does this means it is impossible to get a project having multiple repositories configured as long as they have mirrored jars?

**Comment by ljacomet** [comment]
@Ordiel Could you rephrase your question? I am not sure what you are asking for. Maybe add an example.

**Comment by melix** [comment]
@Ordiel I'm sorry I don't understand your question. What this means is that Gradle does not support fetching a single module which is "spread" over different repositories. So say you have 2 repositories, the first one contains metadata but no jars, and the other one contains jars but no metadata: Gradle will _not_ support this.

However, as @bigdaz explained, if one of your repositories has the complete set (metadata + jars), _then_ you can use [repository filtering](https://docs.gradle.org/current/userguide/userguide_single.html#sec::matching_repositories_to_dependencies) to skip the corrupt repos.
